From: OpenWRT Builder <builder@openwrt.org>
Date: Mon, 13 Oct 2025 16:40:00 +0200
Subject: [PATCH] arm64: dts: qcom: Add MSM8916 MF68E device with ST7735S display

Add device tree for MF68E 4G LTE Mobile WiFi Router with ST7735S display support.

Features:
- ST7735S 128x128 display via SPI0
- PWM backlight via PM8916 MPP4
- Battery monitoring
- GPIO LEDs and buttons
- SPIDEV support for debugging (commented out)

Signed-off-by: OpenWRT Builder <builder@openwrt.org>
---
 arch/arm64/boot/dts/qcom/Makefile             |   1 +
 .../dts/qcom/msm8916-generic-mf68e.dts        | 257 ++++++++++++++++++
 2 files changed, 258 insertions(+)
 create mode 100644 arch/arm64/boot/dts/qcom/msm8916-generic-mf68e.dts

--- a/arch/arm64/boot/dts/qcom/Makefile
+++ b/arch/arm64/boot/dts/qcom/Makefile
@@ -25,6 +25,7 @@ dtb-$(CONFIG_ARCH_QCOM)	+= msm8916-alcatel-idol347.dtb
 dtb-$(CONFIG_ARCH_QCOM)	+= msm8916-asus-z00l.dtb
 dtb-$(CONFIG_ARCH_QCOM)	+= msm8916-huawei-g7.dtb
 dtb-$(CONFIG_ARCH_QCOM)	+= msm8916-longcheer-l8150.dtb
+dtb-$(CONFIG_ARCH_QCOM)	+= msm8916-generic-mf68e.dtb
 dtb-$(CONFIG_ARCH_QCOM)	+= msm8916-mtp.dtb
 dtb-$(CONFIG_ARCH_QCOM)	+= msm8916-samsung-a3u-eur.dtb
 dtb-$(CONFIG_ARCH_QCOM)	+= msm8916-samsung-a5u-eur.dtb
--- /dev/null
+++ b/arch/arm64/boot/dts/qcom/msm8916-generic-mf68e.dts
@@ -0,0 +1,257 @@
+// SPDX-License-Identifier: GPL-2.0-only
+
+/dts-v1/;
+
+#include "msm8916-ufi.dtsi"
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/leds/common.h>
+
+/ {
+	model = "MF68E 4G LTE Mobile WiFi Router";
+	compatible = "generic,mf68e", "qcom,msm8916";
+
+	bat: battery {
+		compatible = "simple-battery";
+		voltage-min-design-microvolt = <3400000>;
+		voltage-max-design-microvolt = <4350000>;
+		energy-full-design-microwatt-hours = <11100000>;
+		charge-full-design-microamp-hours = <3000000>;
+
+		ocv-capacity-celsius = <25>;
+		ocv-capacity-table-0 = <4330000 100>, <4265000 95>,
+			<4208000 90>, <4154000 85>, <4102000 80>,
+			<4063000 75>, <4006000 70>, <3972000 65>,
+			<3931000 60>, <3880000 55>, <3841000 50>,
+			<3817000 45>, <3798000 40>, <3783000 35>,
+			<3767000 30>, <3752000 25>, <3732000 20>,
+			<3712000 16>, <3694000 13>, <3681000 11>,
+			<3675000 10>, <3673000 9>, <3664000 8>,
+			<3655000 7>, <3642000 6>, <3628000 5>,
+			<3611000 4>, <3573000 3>, <3531000 2>,
+			<3495000 1>, <3400000 0>;
+	};
+
+	backlight: backlight {
+		compatible = "pwm-backlight";
+		pwms = <&pm8916_pwm 0 100000>;
+		brightness-levels = <0 255>;
+		num-interpolated-steps = <255>;
+		default-brightness-level = <128>;
+	};
+
+	gpio-keys {
+		compatible = "gpio-keys";
+
+		pinctrl-0 = <&button_default>;
+		pinctrl-names = "default";
+
+		label = "GPIO Buttons";
+
+		button-reset {
+			label = "Reset";
+			gpios = <&tlmm 37 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+		};
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		pinctrl-0 = <&gpio_leds_default>;
+		pinctrl-names = "default";
+
+		led-0 {
+			gpios = <&tlmm 20 GPIO_ACTIVE_HIGH>;
+			color = <LED_COLOR_ID_BLUE>;
+			function = LED_FUNCTION_WLAN;
+			default-state = "off";
+		};
+
+		led-1 {
+			gpios = <&tlmm 21 GPIO_ACTIVE_HIGH>;
+			color = <LED_COLOR_ID_BLUE>;
+			function = LED_FUNCTION_WAN;
+			default-state = "off";
+		};
+
+		led-2 {
+			gpios = <&tlmm 22 GPIO_ACTIVE_HIGH>;
+			color = <LED_COLOR_ID_GREEN>;
+			function = LED_FUNCTION_POWER;
+			default-state = "on";
+		};
+	};
+};
+
+&battery {
+	monitored-battery = <&bat>;
+};
+
+&button_default {
+	pins = "gpio37";
+	function = "gpio";
+	drive-strength = <2>;
+	bias-pull-up;
+};
+
+&gpio_leds_default {
+	pins = "gpio20", "gpio21", "gpio22";
+	function = "gpio";
+	drive-strength = <2>;
+	bias-disable;
+};
+
+&mpss_mem {
+	reg = <0x0 0x86800000 0x0 0x5500000>;
+};
+
+&pm8916_pwm {
+	pinctrl-0 = <&pwm_out>;
+	pinctrl-names = "default";
+	status = "okay";
+};
+
+&pm8916_codec {
+	qcom,micbias-lvl = <2800>;
+	qcom,mbhc-vthreshold-low = <75 150 237 450 500>;
+	qcom,mbhc-vthreshold-high = <75 150 237 450 500>;
+	qcom,hphl-jack-type-normally-open;
+};
+
+&sdhc_1 {
+	status = "okay";
+};
+
+&sdhc_2 {
+	status = "okay";
+	non-removable;
+};
+
+&sound {
+	audio-routing =
+		"AMIC1", "MIC BIAS Internal1",
+		"AMIC2", "MIC BIAS External1",
+		"AMIC3", "MIC BIAS External1";
+};
+
+&usb {
+	dr_mode = "peripheral";
+	extcon = <&usb_id>;
+};
+
+&usb_hs_phy {
+	extcon = <&usb_id>;
+};
+
+&wcnss {
+	status = "okay";
+};
+
+&wcnss_iris {
+	compatible = "qcom,wcn3620";
+};
+
+&tlmm {
+	/*
+	 * GPIO 0,1,2,3: SPI0 for ST7735S display
+	 * GPIO 23: DC (Data/Command) for display
+	 * GPIO 25: Reset for display
+	 */
+	spi0_default: spi0-default-state {
+		pins = "gpio0", "gpio1", "gpio3";
+		function = "blsp_spi1";
+		drive-strength = <8>;
+		bias-disable;
+	};
+
+	spi0_cs_default: spi0-cs-default-state {
+		pins = "gpio2";
+		function = "gpio";
+		drive-strength = <8>;
+		bias-disable;
+		output-high;
+	};
+
+	display_pins: display-state {
+		reset-pins {
+			pins = "gpio25";
+			function = "gpio";
+			drive-strength = <8>;
+			bias-disable;
+			output-high;
+		};
+
+		dc-pins {
+			pins = "gpio23";
+			function = "gpio";
+			drive-strength = <8>;
+			bias-disable;
+			output-high;
+		};
+	};
+};
+
+&pm8916_mpps {
+	/*
+	 * MPP4 is used for PWM backlight control
+	 * Configured for digital output to drive PWM
+	 */
+	pwm_out: mpp4-state {
+		pins = "mpp4";
+		function = "digital";
+		power-source = <PM8916_MPP_VPH>;
+		output-low;
+		qcom,dtest = <1>;
+	};
+};
+
+/*
+ * SPI0 Bus - ST7735S Display Controller
+ * Physical panel: 132x132 pixels
+ * Visible area: 128x128 pixels
+ * Interface: 4-wire SPI (MOSI, MISO, SCK, CS)
+ * Additional: DC (GPIO23), Reset (GPIO25)
+ */
+&spi0 {
+	status = "okay";
+	pinctrl-0 = <&spi0_default &spi0_cs_default &display_pins>;
+	pinctrl-names = "default";
+
+	cs-gpios = <&tlmm 2 GPIO_ACTIVE_LOW>;
+
+	/* display@0 {
+		compatible = "sitronix,st7735s";
+		reg = <0>;
+		spi-max-frequency = <16000000>;
+
+		reset-gpios = <&tlmm 25 GPIO_ACTIVE_HIGH>;
+		dc-gpios = <&tlmm 23 GPIO_ACTIVE_HIGH>;
+
+		/*
+		 * Power supplies from PM8916
+		 * L17: 2.85V for display VDD
+		 * L6: 1.8V for display VDDIO
+		 */
+		VDD-supply = <&pm8916_l17>;
+		VDDIO-supply = <&pm8916_l6>;
+
+		backlight = <&backlight>;
+		rotation = <0>;
+	};*/
+
+	/*
+	 * SPIDEV for debugging - uncomment to enable direct SPI access
+	 * WARNING: Comment out display@0 node above when using spidev
+	 *
+	 * Usage:
+	 *   ls /dev/spidev0.0
+	 *   spi-tools test /dev/spidev0.0
+	 */
+	
+	spidev@0 {
+		compatible = "rohm,dh2228fv";
+		reg = <0>;
+		spi-max-frequency = <1000000>;
+	};
+	
+};
